autorest/azure/async.go

```diff
@@ -37,6 +38,80 @@ const (
 	operationSucceeded  string = "Succeeded"
 )
 
+// Future provides a mechanism to access the status and results of an asynchronous request.
+// Since futures are stateful they should be passed by value to avoid race conditions.
+type Future struct {
+	req  *http.Request
+	resp *http.Response
+	ps   pollingState
+}
+
+// NewFuture returns a new Future object initialized with the specified request.
+func NewFuture(req *http.Request) Future {
+	return Future{req: req}
+}
+
+// Response returns the last HTTP response or nil if there isn't one.
+func (f Future) Response() *http.Response {
+	return f.resp
+}
+
+// Status returns the last status message of the operation.
+func (f Future) Status() string {
+	if f.ps.State == "" {
+		return "Unknown"
+	}
+	return f.ps.State
+}
+
+// Done queries the service to see if the operation has completed.
+func (f *Future) Done(sender autorest.Sender) (bool, error) {
+	// exit early if this future has terminated
+	if f.ps.hasTerminated() {
+		return f.terminalReturn()
+	}
+
+	resp, err := sender.Do(f.req)
+	f.resp = resp
+	if err != nil {
+		return false, err
+	}
+
+	err = updatePollingState(resp, &f.ps)
+	if err != nil {
+		return false, err
+	}
+
+	if f.ps.hasTerminated() {
+		return f.terminalReturn()
+	}
+
+	f.req, err = newPollingRequest(f.ps)
+	return false, err
+}
+
+func (f *Future) terminalReturn() (bool, error) {
+	if !f.ps.hasSucceeded() {
+		return true, f.ps
+	}
+	return true, nil
```

Comment:
I'd a comment here mentioning that this being `true` is intentional. It make sense if you read this in combination with `Done` but otherwise it could be confusing.

Reply:
Good point I'll add a comment.